Implement rendezvous protocol spec

vyzo opened the pull request:
Implementation of spec in https://github.com/libp2p/specs/pull/44

Provides Client, Service, and Database for attaching and using rendezvous points in libp2p hosts and the necessary hooks for implementing federated rendezvous daemons out of package.

TBD:
- [ ] docstrings
- [ ] vendor sqlite in gx
- [ ] README
- [ ] CI

vyzo committed:
rendezvous protobuf

vyzo committed:
client implementation

vyzo committed:
update client interface

It's DiscoverOnce and Discover instead of Discover and DiscoverAsync

vyzo committed:
more fine-grained rendezvous api

vyzo committed:
add error notification for background register/discover

vyzo committed:
interface ergonomics

vyzo committed:
add E_INVALID_TTL to rendezvous.proto

vyzo committed:
include namespace in error logs

vyzo committed:
simplify Rendezvous interface

vyzo committed:
use discriminated registration errors

vyzo committed:
annotate registration error

vyzo committed:
update protobuf

vyzo committed:
update for response error changes in the protocol

vyzo committed:
rendezvous interface should expose full registration information

- instead of returning PeerInfos, return Registration objects in discovery
- provide utility functions at module level for peer info discovery

vyzo committed:
service implementation

vyzo committed:
rendezvous service sync: hook for federation

vyzo committed:
Registration records should have an actual peer ID

vyzo committed:
client: use larger batch in discovery, only poll immediately if full

vyzo committed:
client: add TODO comment for adaptive backoff

vyzo committed:
refactor service constructor into two parts

vyzo committed:
user parameter for ttl in Register

vyzo committed:
database logic implementation

vyzo committed:
implement binary packing details

vyzo committed:
better logging for service i/o

vyzo committed:
test address and cookie packing

vyzo committed:
test db functionality

vyzo committed:
up MaxRegistrations to 1k

vyzo committed:
test db functionality with multiple namespaces

vyzo committed:
basic service test

vyzo committed:
test service errors

vyzo committed:
make db nonce 32 bytes

vyzo committed:
test client specific functionality

vyzo commented:
Summoning @whyrusleeping @Stebalien 
CC @magik6k @dryajov @diasdavid

magik6k commented:
Why do we need sqlite here and can't just use the datastore?

vyzo commented:
@magik6k it does queries that are much easier to express and manage in SQLite than the datastore.

magik6k commented:
On CI - assuming `ci` team on github has access to this repo, you can enable jenkins by just creating `ci/Jenkinsfile` with `golang()` as content - like https://github.com/ipfs/ipfs-ds-convert/blob/master/ci/Jenkinsfile

vyzo commented:
Sure, let's have both travis and jenkins.

victorb commented:
@vyzo think you can skip Travis, it's not giving us a lot of value right now and mostly used as a check that existing repositories gives the same status on Travis as on Jenkins.

vyzo commented:
Sure, didn't know we were sunsetting travis; less work to do :)

vyzo committed:
use randomized exponential backoff in error retry for persistent client registrations

vyzo commented:
Finetuned the behaviour of persistent client registrations on refresh errors; it now uses exponential backoff instead of blindly waiting for another ttl.

vyzo committed:
client: add TODO for robust discovery error recovery

whyrusleeping commented on client.go:
```diff
@@ -0,0 +1,287 @@
+	DiscoverAsyncInterval = 2 * time.Minute
+)
+
+type Rendezvous interface {
```
comments on the interface would be helpful

vyzo replied:
yes indeed! Part of the "docstrings" checkbox.

whyrusleeping commented on client.go:
```diff
@@ -0,0 +1,287 @@
+	return nil
+}
+
+func registerRefresh(ctx context.Context, rz Rendezvous, ns string, ttl int) {
```
It doesnt appear that this gets closed down when unregister is called. So if i call register then unregister, this routine would re-register me.

vyzo replied:
hrm, yes, the two are not correlated; you'd have to cancel the context to stop registration refresh.

vyzo replied:
Note that this is a utility function to augment the interface with persistent registrations.

vyzo replied:
Maybe we should make a more complicated stateful client contraption that handles the registration/unregistration correlation?

I think we can have two interfaces, a low level `RendezvousPoint` interface that has single stateless operations and a high level `Rendezvous` interface that incorporates the module level utility functions.

whyrusleeping commented on client.go:
```diff
@@ -0,0 +1,287 @@
+		}
+
+		if len(regs) < batch {
+			// TODO adaptive backoff for heavily loaded rendezvous points
```
Can the service respond something like "load too high, try again in a bit" ? Seems like a nice DoS mitigation feature.

vyzo replied:
Not sure if a load indicator would help, as the clients can simply ignore it.
I am in favour of load inference from the rate of registration responses.

vyzo replied:
On the other hand, we could make this an Error response -- `E_TEMPORARILY_UNAVAILABLE` perhaps.

Now that could help, as it would force the clients to back off and possibly create a new stream.

whyrusleeping replied:
Yeah, I was thinking have it be an error response. So they don't get any records back, and the error tells them to just wait and try again in a bit

whyrusleeping commented on db.go:
```diff
@@ -0,0 +1,452 @@
+	"os"
+	"time"
+
+	_ "github.com/mattn/go-sqlite3"
```
probably good to have the client and server code separated out. I'd rather not import sqlite in my program if i don't have to

vyzo replied:
Probably a good idea. 

Should I make a separate `go-libp2p-rendezvous-service` package?

We might be able to get away with simply putting it in a subpackage though, so that unless you import `github.com/go-libp2p-rendezvous/svc` (or however we call it), you shouldn't get the db code.

whyrusleeping replied:
yeah, subpackage is fine for now i think

vyzo replied:
ok, will restructure.

vyzo replied:
Restructured in https://github.com/libp2p/go-libp2p-rendezvous/pull/1/commits/c487c2077b1528816577f043c1f56d143915a63b

We now have a db subpackage with the interface, and implementations of the db in subpackages of that.

So now there is no dependency of the main package (or the dbi interface package) on SQLite or any particular implementation, and clients can freely import the client interface without needing to link SQLite code.

whyrusleeping commented on db.go:
```diff
@@ -0,0 +1,452 @@
+)
+
+type DB struct {
+	db *sql.DB
```
i'm happy to have this be sqlite as long as its just an implementation detail (anyone could implement the service using any other form of backing datastore)

vyzo replied:
Yeah, it's totally a service implementation detail.

vyzo replied:
After the restructuring in https://github.com/libp2p/go-libp2p-rendezvous/commit/c487c2077b1528816577f043c1f56d143915a63b the implementation of the database is hidden behind an opaque interface.

So we can have implementations using different datastores etc; up to the user to choose the database backing the service.

whyrusleeping commented on svc.go:
```diff
@@ -0,0 +1,244 @@
+	rcount, err := rz.DB.CountRegistrations(p)
+	if err != nil {
+		log.Errorf("Error counting registrations: %s", err.Error())
+		return newRegisterResponseError(pb.Message_E_INTERNAL_ERROR, err.Error())
```
including raw text from a sql error in a response to the client seems like a potentially bad information leak. Just a gut feeling though.

vyzo replied:
Uhm, you are probably right -- I'll neuter this.

vyzo replied:
Plugged the leaks in https://github.com/libp2p/go-libp2p-rendezvous/pull/1/commits/baf1e4e61812c08f5e25ac019e04cbabdced4ce4

It's just a "database error" to the client now.

whyrusleeping commented:
does the service do period sweeps to clean up expired entries? I couldnt find where that was happening if it is

whyrusleeping commented:
I'd also love to get some metrics plugged in here so we can observe it more easily

vyzo commented:
> does the service do period sweeps to clean up expired entries? I couldnt find where that was happening if it is

Yes, it is in the db itself -- See `db.background` and `db.cleanupExpired`

vyzo commented:
> I'd also love to get some metrics plugged in here so we can observe it more easily

Sure, that's a good suggestion. What should I utilize for this?

whyrusleeping commented:
Exposing metrics via prometheus is how we usually do it. Ping @Kubuxu and @lgierth for input here

vyzo committed:
refactor database interface and implementation into db subpackage

We now have:
- db/dbi.go -- interface, no dependency on any particular database
- db/sqlite/... -- sqlite implementation, depends on dbi
- the rendezvous.NewRendezvousService API has been redefined to NewRendezvouServiceWithDB
  service importer constructs the database of choice
- the rendezvous package has no dependences on db implementation, so clients can import
  without linking SQLite in the binary

vyzo committed:
don't leak database error details in internal errors

vyzo committed:
two interfaces for client-side: RendezvousPoint and RendezvousClient

RendezvousClient soon to be stateful.

vyzo commented:
Pushed WIP stateful client so that I don't lose it or forget about it.
Further work:
- Flesh out the stateful client so that it consistently handles registrations and caches discovery results for quick, randomized access to the peer set.
- add rate limiter to the service, with an `E_UNAVAILABLE` status code prompting client backoff.
- add metrics collection to the service

Kubuxu commented:
For metrics use https://github.com/ipfs/go-metrics-interface, go-ipfs by default installs https://github.com/ipfs/go-metrics-prometheus to expose it using Prometheus protocol.

bigs commented:
@whyrusleeping @vyzo how's this issue doing? i'd be happy to review it thursday.

vyzo commented:
Review would be nice, I plan to restart work on it soonish -- see the further work comment.

vyzo committed:
update protobuf

vyzo committed:
include ttl in registration response

vyzo committed:
update gx deps

vyzo committed:
expose counter in register interface

vyzo committed:
update tests

raulk commented:
Hey @vyzo – wondering what you have in mind as a path forward here. Is this implementation already 100% compliant with our [flashy new Rendezvous spec](https://github.com/libp2p/specs/tree/master/rendezvous)?

vyzo commented:
@raulk yes, it's 100% compatible with the spec. It needs rate limiting to emit the `E_UNAVAILABLE` response though and I need to figure out how to do federation (although the federated daemon will be a separate package).

raulk commented:
@vyzo — cool. How about we figure out the error case and we merge this in marking it as EXPERIMENTAL? No point keeping it parked in a PR. We can then keep iterating on master.

vyzo commented:
I also want to work a bit on the stateful client, but we should be able to merge soon.

vyzo commented:
The federation can wait for the out of package daemon implementation.

vyzo commented:
I'll work a bit more on it this week.

mikedoan commented:
@vyzo when are you expecting this will be checked in?

vyzo commented:
soon [TM]

aschmahmann committed:
Switched from gx to go mod and started using go-libp2p-core interfaces

aschmahmann committed:
Add stateful discovery client

aschmahmann committed:
RendezvousPoint and RendezvousClient now return the server's TTL on Registered records.
Default TTL for discovery client increased
discovery client now utilizes server cookie for added efficiency

aschmahmann committed:
fixed compile error from previous commit + code refactoring

aschmahmann committed:
replaced sync.Map with map + RW mutex. small refactors

vyzo committed:
Merge pull request #3 from aschmahmann/feat/add-discovery-client

Feat/add discovery client

10d9e commented:
Is this in a place to be reviewed/merged @Stebalien @Kubuxu?

mxinden commented:
For anyone continuing this pull request, please note that the specification had [multiple revisions](https://github.com/libp2p/specs/commits/master/rendezvous/README.md) in the meantime.
